php/libraries/FileUploadValidator.class.inc

```diff
@@ -0,0 +1,241 @@
+                $this->_responseCode = 409;
+                return false;
+            }
+            // Validate that the MIME type of the file is allowed
+            if (!empty($this->permittedMIMETypes)) {
+                if (!in_array(
+                    $uploadedFile->getClientMediaType(),
+                    $this->permittedMIMETypes,
+                    true
+                )
+                ) {
+                    $this->_responseCode = 400;
+                    return false;
+                }
+            }
+
+            // Check that the file size does not exceed the upload limit
+            $limit = $this->fileUploadMaxSize();
+            if ($uploadedFile->getSize() > $limit) {
+                error_log(
+                    "Uploaded file size {$uploadedFile->getSize} exceeds "
+                    . "limit of $limit"
+                );
+                $this->_responseCode = 400;
+                return false;
+            }
+
+            // Move the file to $targetPath
+            $uploadedFile->moveTo($targetPath);
+
+            chmod($targetPath, $permissions);
+        }
+
+        // If all goes well, set response code to 'Created'
+        $this->_responseCode = 201;
+        return true;
+    }
+
+    /**
+     * Returns the value of private int $_responseCode
+     *
+     * @return int The HTTP repsonse code corresponding to the upload status.
+     */
+    public function getResponseCode(): int
+    {
+        return $this->_responseCode;
+    }
+
+    /**
+     * Returns a file size limit in bytes based on the PHP upload_max_filesize
+     * and post_max_size.
+     *
+     * @return int The smaller of PHP's upload_max_filesize and post_max_size
+     *                  settings.
+     */
+    function fileUploadMaxSize(): int
+    {
+        static $max_size = -1;
+
+        if ($max_size < 0) {
+            // Start with post_max_size.
+            $post_max_size = $this->parseSize(ini_get('post_max_size'));
+            if ($post_max_size > 0) {
+                $max_size = $post_max_size;
+            }
+
+            // If upload_max_size is less, then reduce. Except if upload_max_size is
+            // zero, which indicates no limit.
+            $upload_max = $this->parseSize(ini_get('upload_max_filesize'));
+            if ($upload_max > 0 && $upload_max < $max_size) {
+                $max_size = $upload_max;
+            }
+        }
+        return $max_size;
+    }
+
+    /**
+     * Parse a PHP size string such as "4G" into an integer equivalent to the
+     * number of bytes that string represents.
+     *
+     * @param string $size A string representing an amount of memory.
+     *
+     * @return int An integer corresponding to the number of bytes in the
+     *              byte string
+     */
+    function parseSize(string $size): int
+    {
+        // Remove the non-unit characters from the size.
+        $unit = preg_replace('/[^bkmgtpezy]/i', '', $size);
```

Comment:
if this line was `$unit = preg_replace('/[^bkmgtpezy]/i', '', $size)[0] ?? '';` the `if` statement could be removed and always return `return round($size * pow(1024, stripos('bkmgtpezy', $unit)));`